File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/cpp/segment-tree.cpp
/**
 * C++ Implementation: Segment Tree
 * Invariant: Range Query O(log N), Point/Range Update O(log N)
 */
#include <iostream>
#include <vector>
#include <stdexcept>

template <typename T>
class SegmentTree {
private:
    std::vector<T> elements;
public:
    // TODO: Implement primary invariant operations
    void push(const T& val) {
        elements.push_back(val);
    }
    size_t size() const { return elements.size(); }
};

File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/java/segment-tree.java
/**
 * Java Implementation: Segment Tree
 * Invariant: Range Query O(log N), Point/Range Update O(log N)
 */
package seed.dsa.core;

import java.util.*;

public class SegmentTree<T> {
    private final List<T> items = new ArrayList<>();

    // TODO: Implement your solution here
    public void add(T value) {
        items.add(value);
    }

    public int size() {
        return items.size();
    }
}
